Fix side-effect check on loop invariants

SaswatPadhi opened the pull request:
Fixes #5915.

Invariant predicates were being _cleaned_ (C -> logic via `clean_expr`) within `convert_loop_invariant` and were rejected if any intermediate goto instructions were introduced during the cleaning process. However, this check is overly conservative -- it ends up rejecting some complex, but pure, Boolean expressions that are compiled to multiple goto instructions.

We replace this logic with a call to `has_subexpr` with `ID_side_effect`.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- N/A ~~The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/~~
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- N/A ~~My commit message includes data points confirming performance improvements (if claimed).~~
- [x] My PR is restricted to a single feature or bugfix.
- N/A ~~White-space or formatting changes outside the feature-related changed lines are in commits of their own.~~

martin-cs commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -868,13 +868,11 @@ void goto_convertt::convert_loop_invariant(
+  // at multiple code points around a loop.
+  // Rather than performing `clean_expr` during compilation,
+  // we convert the invariant to logic at these various code points
+  // only when they are enforced (using goto-instrument).
```
So... this means that the output of goto_convert will no longer be side effect free?  So I can now write code that has a different meaning if run with and without code contracts?  That doesn't seem like a very safe design.

tautschnig replied:
Without having reviewed the details just yet: the short term idea is to move all invariant-related processing (as far as possible) to `code_contracts.cpp`, knowing that this will leave annotations in there that wouldn't satisfy side-effect freedom. Once this has been cleaned up (right now it's a mess with parts of the processing happening in all sorts of places), it should be fully integrated into goto-program conversion, possibly leaving little or (ideally) nothing left in `goto-instrument`.

SaswatPadhi committed:
Fix side-effect check on loop invariants

Loop invariants were rejected during compilation if any intermediate goto
instructions were introduced during the cleaning process.
However, this check is overly conservative -- it ends up rejecting some complex,
but pure, Boolean expressions that are compiled to multiple goto instructions.

We replace this logic with a call to `has_subexpr` with `ID_side_effect`.

SaswatPadhi committed:
Minor refactor and cleanup on code contracts

We can maintain a `protected` `goto_convertt` field in our class, which would be reusable, rather than trying to instantiate an object for each codet -> goto_programt transformation.

SaswatPadhi commented:
@martin-cs @tautschnig 

After a lot of back and forth discussion and changes, it seemed like at least for the invariants that we currently write, it's sufficient to just check for `ID_side_effect` subexpressions. I left a TODO to revisit the `has_sideeffect` function later and allow calling functions that have an empty ASSIGNS clause from within loop invariants. If you think we might also need to handle some other cases, please suggest -- I'd be happy to include them as part of this change, or leave a TODO to revisit later (if it gets too complicated).

tautschnig commented on src/goto-programs/goto_convert_side_effect.cpp:
```diff
@@ -35,6 +35,24 @@ bool goto_convertt::has_function_call(const exprt &expr)
   return false;
 }
 
+bool goto_convertt::has_sideeffect(const exprt &expr)
+{
+  // TODO: In the future, we want to allow some function calls,
+  // specifically those which have an empty ASSIGNS clause annotation.
+  if(expr.id() == ID_side_effect)
+  {
+    return true;
+  }
+
+  for(const auto &op : expr.operands())
+  {
+    if(goto_convertt::has_sideeffect(op))
+      return true;
+  }
+
+  return false;
+}
+
```
Good call out that this was never implemented. That said, can you actually do this a bit different as follows:
1) Get rid of the declaration as well.
2) At the point where you will currently need it, just use `has_subexpr(expr, ID_side_effect);` (from `util/expr_util.h`).

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -112,17 +112,20 @@ static void check_apply_invariants(
-      goto_programt::make_assertion(invariant, loop_head->source_location));
-    a->source_location.set_comment("Check loop invariant before entry");
+    auto assertion = code_assertt(invariant);
+    assertion.add_source_location().swap(loop_head->source_location);
```
Is `swap` the right thing to do here?

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -112,17 +112,20 @@ static void check_apply_invariants(
-    goto_programt::targett a = havoc_code.add(
-      goto_programt::make_assertion(invariant, loop_head->source_location));
-    a->source_location.set_comment("Check loop invariant before entry");
+    auto assertion = code_assertt(invariant);
```
Nit picking: is this use of `auto` adding value? Wouldn't `code_assertt assertion{invariant};` be even fewer characters and thus less cognitive load?

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -112,17 +112,20 @@ static void check_apply_invariants(
   // assume the invariant
-  havoc_code.add(
-    goto_programt::make_assumption(invariant, loop_head->source_location));
+  auto assumption = code_assumet(invariant);
```
As above: does `auto` add value?

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -138,11 +141,14 @@ static void check_apply_invariants(
-    a.source_location.set_comment("Check that loop invariant is preserved");
-    goto_function.body.insert_before_swap(loop_end, a);
-    ++loop_end;
+    auto assertion = code_assertt(invariant);
```
Does `auto` add value?

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -112,17 +112,20 @@ static void check_apply_invariants(
-  havoc_code.add(
-    goto_programt::make_assumption(invariant, loop_head->source_location));
+  auto assumption = code_assumet(invariant);
+  assumption.add_source_location().swap(loop_head->source_location);
```
Given the earlier use of `swap`, there won't be any source location in here?

tautschnig commented on src/goto-instrument/code_contracts.cpp:
```diff
@@ -138,11 +141,14 @@ static void check_apply_invariants(
-    goto_function.body.insert_before_swap(loop_end, a);
-    ++loop_end;
+    auto assertion = code_assertt(invariant);
+    assertion.add_source_location().swap(loop_end->source_location);
```
As above: nothing left to swap in.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -18,6 +18,8 @@ Date: February 2016
 #include <string>
 #include <unordered_set>
 
+#include <analyses/local_may_alias.h>
```
No need to have this in the header file, a forward declaration will do.

tautschnig commented on src/goto-instrument/code_contracts.h:
```diff
@@ -26,6 +28,8 @@ Date: February 2016
 #include <util/namespace.h>
 #include <util/pointer_expr.h>
 
+#include "loop_utils.h"
```
Wouldn't forward declarations do the trick?

SaswatPadhi replied:
Sorry, resolved this comment by mistake without addressing it.

I `#include`d this header file for the `loopt` type, which is `typedef`ed as:

```c
typedef const natural_loops_mutablet::natural_loopt loopt;
```

Could I forward declare it here somehow? I wasn't sure how to.

tautschnig replied:
I believe one of `typename loopt;` or `class loopt;` might still work, and likely the compiler will telly you if it doesn't :-) Not high priority, though.

tautschnig commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -868,13 +868,11 @@ void goto_convertt::convert_loop_invariant(
-    no_sideeffects.instructions.empty(),
-    "loop invariant is not side-effect free",
-    code.find_source_location());
+  if(has_sideeffect(invariant))
```
See comment in the previous commit: use `has_subexpr(invariant, ID_side_effect)`

martin-cs commented on src/goto-programs/goto_convert_class.h:
```diff
@@ -97,7 +97,6 @@ class goto_convertt:public messaget
 
   void rewrite_boolean(exprt &dest);
 
-  static bool has_sideeffect(const exprt &expr);
```
Why is this removed?

SaswatPadhi replied:
It was never defined, so I removed the declaration as well.

martin-cs replied:
Sounds good.

martin-cs commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -868,13 +868,11 @@ void goto_convertt::convert_loop_invariant(
-    no_sideeffects.instructions.empty(),
-    "loop invariant is not side-effect free",
-    code.find_source_location());
+  if(has_subexpr(invariant, ID_side_effect))
```
As I recall, this was the really contentious part last time around.  This seems much less disruptive.

I think throwing the exception is correct because the user can trigger this even if the tool is non-buggy.  I wonder if it would be more user-friendly to check this earlier but ... I guess we need to check it here as well.

I wonder if this might still be a bit too conservative but, if it works for you then fine.

SaswatPadhi replied:
Thank you, yes earlier I had removed this part entirely and was doing some checking in `goto-instrument` instead. Now this check is performed at `goto-cc` compilation.

> I wonder if this might still be a bit too conservative but, if it works for you then fine.

I completely agree with you. And it's on our roadmap to allow function calls within contracts, if those functions have an empty assigns clause (so no side effects).

martin-cs reviewed (approved):
My initial reason for objection, that we were changing a major post condition of goto-convert is gone so there is no reason for me to block this.

Some thoughts:
*. Are similar conditions checked for pre/post?
*. Are the restrictions documented in the manual?

Longer-term, do we want to ban function calls from these annotations?  One thing I found nice about SPARK was being able to define predicates (like `is_initialised` and `is_well_structured`) using Ada functions.  If there isn't a way of doing this you risk a lot of parallel maintenance when altering common invariants.  I guess macros would be one option but these don't work for things like `is_sorted_tree` and having options of how much to expand them.   This is well beyond the scope of this PR but is something worth thinking about IMHO and will touch this code.

SaswatPadhi commented:
Thank you for taking a look, @martin-cs.

I just wanted to mention that @feliperodri and I are actively working on the issues that you pointed out:

> Are similar conditions checked for pre/post?

Not currently, but we are fix this during our refactoring pass over the code contracts module

> Are the restrictions documented in the manual?

Good call out! Nope, they are not. I would add it to our TODO list for documentation updates.

> Longer-term, do we want to ban function calls from these annotations?

It's on our roadmap to allow function calls within contracts, if those functions have an empty assigns clause (so no side effects).
